### PR TITLE
sdl >= 2.0.18: fix CMake options

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **lib/1.0**

closes https://github.com/conan-io/conan-center-index/issues/8607

---

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/sdl/all/conanfile.py [hunk 1]
```diff
@@ -176,19 +176,21 @@ def build(self):
...
+        if self._cmake:
+            return self._cmake
+        self._cmake = CMake(self)
+        cmake_required_includes = []  # List of directories used by CheckIncludeFile (https://cmake.org/cmake/help/latest/module/CheckIncludeFile.html)
+        cmake_extra_ldflags = []
+        # FIXME: self.install_folder not defined? Neccessary?
+        self._cmake.definitions["CONAN_INSTALL_FOLDER"] = self.install_folder
+        if self.settings.os != "Windows" and not self.options.shared:
+            self._cmake.definitions["SDL_STATIC_PIC"] = self.options.fPIC
+        if self.settings.compiler == "Visual Studio" and not self.options.shared:
```
Comment (jgsogo) [inline recipes/sdl/all/conanfile.py:188]:
```suggestion
        if self.settings.compiler in ["Visual Studio", "msvc"] and not self.options.shared:
```

wdyt? Should we start considering `msvc` or maybe it is better to create a _tool_ to check for VS compiler?

File: recipes/sdl/all/conanfile.py [hunk 2]
```diff
@@ -247,18 +249,84 @@ def _configure_cmake(self):
                     # FIXME: Otherwise 2.0.16 links with system wayland (from egl/system requirement)
                     cmake_extra_ldflags += ["-L{}".format(os.path.join(self.deps_cpp_info["wayland"].rootpath, it)) for it in self.deps_cpp_info["wayland"].libdirs]
                     self._cmake.definitions["WAYLAND_SHARED"] = self.options["wayland"].shared
-                    self._cmake.definitions["WAYLAND_SCANNER_1_15_FOUND"] = 1  # FIXME: Check actual build-requires version
+                    self._cmake.definitions["WAYLAND_SCANNER_1_15_FOUND"] = tools.Version(self.deps_cpp_info["wayland"].version) >= "1.15.0"
```
Comment (jgsogo) [inline recipes/sdl/all/conanfile.py:252]:
```suggestion
                    self._cmake.definitions["WAYLAND_SCANNER_1_15_FOUND"] = 1  # FIXME: Check actual build-requires version
```

using `deps_cpp_info` we get the information from the _host_ context, not the _build_ one. Here we need to check the version of the build-requires, not the requires

File: recipes/sdl/all/conanfile.py [hunk 3]
```diff
@@ -247,18 +249,84 @@ def _configure_cmake(self):
...
+                    self._cmake.definitions["HAVE_XSHAPE_H"] = True
+                self._cmake.definitions["SDL_X11_XVM"] = self.options.xvm
+                if self.options.xvm:
+                    self._cmake.definitions["HAVE_XF86VM_H"] = True
+                self._cmake.definitions["SDL_WAYLAND"] = self.options.wayland
+                if self.options.wayland:
+                    # FIXME: Otherwise 2.0.16 links with system wayland (from egl/system requirement)
+                    cmake_extra_ldflags += ["-L{}".format(os.path.join(self.deps_cpp_info["wayland"].rootpath, it)) for it in self.deps_cpp_info["wayland"].libdirs]
+                    self._cmake.definitions["SDL_WAYLAND_SHARED"] = self.options["wayland"].shared
+                    self._cmake.definitions["WAYLAND_SCANNER_1_15_FOUND"] = tools.Version(self.deps_cpp_info["wayland"].version) >= "1.15.0"
```
Comment (jgsogo) [inline recipes/sdl/all/conanfile.py:318]:
```suggestion
                    self._cmake.definitions["WAYLAND_SCANNER_1_15_FOUND"] = 1  # FIXME: Check actual build-requires version
```

Probably here it also want to check the version from the _build_ context.

**Review by jgsogo — CHANGES_REQUESTED** [review]
We can check if the new model with `self.dependencies` is able to retrieve information from the _build_ context, but we cannot use `deps_cpp_info`.

Probably we can just assume `WAYLAND_SCANNER_1_15_FOUND=1` because the build-requires (`wayland/1.19.0`) is not going to be overriden.

File: recipes/sdl/all/conanfile.py (hunk 2 above)
Comment (SpaceIm) [reply recipes/sdl/all/conanfile.py:252]: wayland is not build context, the comment was wrong

File: recipes/sdl/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/sdl/all/conanfile.py:188]: yes, there are 2 occurrences of Visual Studio, so it shouldn't be a big change.

File: recipes/sdl/all/conanfile.py (hunk 2 above)
Comment (SpaceIm) [reply recipes/sdl/all/conanfile.py:252]: I'll revert this change, didn't see that wayland was also in build requirements.

**Review by jgsogo — APPROVED** [review]

**Review by madebr — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**